subsys/net/ip/net_shell.c

```diff
@@ -3534,159 +3753,249 @@ static void iface_vlan_cb(struct net_if *iface, void *user_data)
-		"\n\tPrint information about network connections" },
-	{ "dns", net_shell_cmd_dns, "\n\tShow how DNS is configured\n"
-		"dns cancel\n\tCancel all pending requests\n"
-		"dns <hostname> [A or AAAA]\n\tQuery IPv4 address (default) or "
-		"IPv6 address for a  host name" },
-	{ "gptp", net_shell_cmd_gptp,
-		"\n\tPrint information about gPTP support\n"
-		"gptp <port>\n\tPrint detailed information about gPTP port" },
-	{ "http", net_shell_cmd_http,
-		"\n\tPrint information about active HTTP connections\n"
-		"http monitor\n\tStart monitoring HTTP connections\n"
-		"http\n\tTurn off HTTP connection monitoring" },
-	{ "iface", net_shell_cmd_iface,
-		"\n\tPrint information about network interfaces\n"
-		"iface up [idx]\n\tTake network interface up\n"
-		"iface down [idx]\n\tTake network interface down" },
-	{ "mem", net_shell_cmd_mem,
-		"\n\tPrint information about network memory usage" },
-	{ "nbr", net_shell_cmd_nbr, "\n\tPrint neighbor information\n"
-		"nbr rm <IPv6 address>\n\tRemove neighbor from cache" },
-	{ "ping", net_shell_cmd_ping, "<host>\n\tPing a network host" },
-	{ "route", net_shell_cmd_route, "\n\tShow network route" },
-	{ "rpl", net_shell_cmd_rpl, "\n\tShow RPL mesh routing status" },
-	{ "stacks", net_shell_cmd_stacks,
-		"\n\tShow network stacks information" },
-	{ "stats", net_shell_cmd_stats,
-		"\n\tShow network statistics\n"
-		"stats all\n\tShow network statistics for all network "
-						"interfaces\n"
-		"stats <idx>\n\tShow network statistics for one specific "
-						"network interfaces\n" },
-	{ "tcp", net_shell_cmd_tcp, "connect <ip> port\n\tConnect to TCP peer\n"
-		"tcp send <data>\n\tSend data to peer using TCP\n"
-		"tcp close\n\tClose TCP connection" },
-	{ "vlan", net_shell_cmd_vlan, "\n\tShow VLAN information\n"
-		"vlan add <vlan tag> <interface index>\n"
-		"\tAdd VLAN tag to the network interface\n"
-		"vlan del <vlan tag>\n"
-		"\tDelete VLAN tag from the network interface\n" },
-	{ NULL, NULL, NULL }
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_dns)
+{
+	SHELL_CMD(cancel, NULL, "Cancel all pending requests.",
+		  cmd_net_dns_cancel),
+	SHELL_CMD(query, NULL, "Query IPv4 address (default) or "
+		  "IPv6 address for a host name.",
+		  cmd_net_dns_query),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_gptp)
+{
+	SHELL_CMD(port, NULL, "Print detailed information about gPTP port.",
+		  cmd_net_gptp_port),
+	SHELL_SUBCMD_SET_END
 };
 
-SHELL_REGISTER(NET_SHELL_MODULE, net_commands);
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_http)
+{
+	SHELL_CMD(monitor, NULL, "Start monitoring HTTP connections.",
+		  cmd_net_http_monitor),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_iface)
+{
+	SHELL_CMD(up, NULL, "Take network interface up.", cmd_net_iface_up),
+	SHELL_CMD(down, NULL, "Take network interface down.",
+		  cmd_net_iface_down),
+	SHELL_SUBCMD_SET_END
+};
+
+SHELL_CREATE_STATIC_SUBCMD_SET(net_cmd_nbr)
+{
+	SHELL_CMD(rm, NULL, "Remove neighbor from cache.",
```

Comment:
You are defining help string for this command (and others):
```
SHELL_CMD(rm, NULL, "**Remove neighbor from cache.**",
		  cmd_net_nbr_rm),
```
However in the command handler: **cmd_net_nbr_rm** you are not calling a function resposible for printing help.
In current command implementation when user will call command: 
`net nbr rm -h` or `net nbr rm --help`
help string "Remove neighbor from cache." will not be dispayed by shell.

To be able to see help string you need to add following code at the beginning of command halder:

`if (shell_help_requested(shell)) {
		shell_help_print(shell, opt, opt_len);
		return;
}`

Reply:
In that case I would recommend using shell_cmd_precheck() function:
> Prints help if request and prints error message on wrong argument count.

Reply:
Hint: Function `shell_help_print` prints help for existing command and subcommands 1 level below.

I prefere to add this function to each command handler, for either root command or subcommands.